File: src/main/java/priv/bajdcc/LALR1/grammar/tree/Block.java
package priv.bajdcc.LALR1.grammar.tree;

import priv.bajdcc.LALR1.grammar.codegen.ICodegen;
import priv.bajdcc.LALR1.grammar.runtime.RuntimeInst;
import priv.bajdcc.LALR1.grammar.semantic.ISemanticRecorder;
import priv.bajdcc.LALR1.grammar.tree.closure.IClosureScope;

import java.util.ArrayList;
import java.util.List;

/**
 * 【语义分析】块
 *
 * @author bajdcc
 */
public class Block implements ICommon {

	/**
	 * 语句集合
	 */
	private List<IStmt> stmts = null;

	public Block() {
		stmts = new ArrayList<>();
	}

	public Block(List<IStmt> stmts) {
		this.stmts = stmts;
	}

	public List<IStmt> getStmts() {
		return stmts;
	}

	@Override
	public void analysis(ISemanticRecorder recorder) {
		for (IStmt stmt : stmts) {
			stmt.analysis(recorder);
		}
	}

	@Override
	public void genCode(ICodegen codegen) {
		codegen.genCode(RuntimeInst.iscpi);
		int start, end;
		for (IStmt stmt : stmts) {
			start = codegen.getCodeIndex();
			stmt.genCode(codegen);
			end = codegen.getCodeIndex() - 1;
			if (start <= end) {
				codegen.genDebugInfo(start, end, stmt.toString());
			}
			if (stmt instanceof StmtReturn) {
				StmtReturn ret = (StmtReturn) stmt;
				if (!ret.isYield())
					break;
			}
		}
		codegen.genCode(RuntimeInst.iscpo);
	}

	@Override
	public String toString() {
		return print(new StringBuilder());
	}

	@Override
	public String print(StringBuilder prefix) {
		StringBuilder sb = new StringBuilder();
		sb.append("{");
		sb.append(System.lineSeparator());
		prefix.append("    ");
		for (IStmt stmt : stmts) {
			sb.append(stmt.print(prefix));
			sb.append(System.lineSeparator());
		}
		prefix.delete(0, 4);
		sb.append(prefix.toString()).append("}");
		return sb.toString();
	}

	@Override
	public void addClosure(IClosureScope scope) {
		for (IStmt stmt : stmts) {
			stmt.addClosure(scope);
			if (stmt instanceof StmtReturn) {
				break;
			}
		}
	}
}

File: src/main/kotlin/com/bajdcc/LALR1/grammar/tree/Block.kt
package com.bajdcc.LALR1.grammar.tree

import com.bajdcc.LALR1.grammar.codegen.ICodegen
import com.bajdcc.LALR1.grammar.runtime.RuntimeInst
import com.bajdcc.LALR1.grammar.semantic.ISemanticRecorder
import com.bajdcc.LALR1.grammar.tree.closure.IClosureScope

/**
 * 【语义分析】块
 *
 * @author bajdcc
 */
class Block(var stmts: MutableList<IStmt> = mutableListOf()) : ICommon {

    override fun analysis(recorder: ISemanticRecorder) {
        stmts.forEach { stmt ->
            stmt.analysis(recorder)
        }
    }

    override fun genCode(codegen: ICodegen) {
        codegen.genCode(RuntimeInst.iscpi)
        for (stmt in stmts) {
            val start = codegen.codeIndex
            stmt.genCode(codegen)
            val end = codegen.codeIndex - 1
            if (start <= end) {
                codegen.genDebugInfo(start, end, stmt.toString())
            }
            if (stmt is StmtReturn) {
                if (!stmt.isYield)
                    break
            }
        }
        codegen.genCode(RuntimeInst.iscpo)
    }

    override fun toString(): String {
        return print(StringBuilder())
    }

    override fun print(prefix: StringBuilder): String {
        val sb = StringBuilder()
        sb.append("{")
        sb.append(System.lineSeparator())
        prefix.append("    ")
        stmts.forEach { stmt ->
            sb.append(stmt.print(prefix))
            sb.append(System.lineSeparator())
        }
        prefix.delete(0, 4)
        sb.append(prefix.toString()).append("}")
        return sb.toString()
    }

    override fun addClosure(scope: IClosureScope) {
        for (stmt in stmts) {
            stmt.addClosure(scope)
            if (stmt is StmtReturn) {
                break
            }
        }
    }
}
